spec/GraphQLOverHTTP.md

```diff
@@ -738,10 +735,15 @@ Note: The GraphQL specification
 and refers to the situation wherein a _GraphQL field error_ occurs as a partial
 response; it still indicates successful execution.
 
-## Processing the response
+## Processing a response
+
+In a typical environment, the source of a response may be some intermediary
+server such as an API gateway, a proxy, a firewall, etc.
+
+Those intermediary servers may not understand GraphQL: in the case of an error,
+they may return their own non-GraphQL `application/json` response with a
+non-`2xx` status code.
```

Comment:
I don't see the relevancy of whether the intermediary servers understand GraphQL.  Perhaps we could say that they 'may not return a GraphQL response', but then, what servers would?  Ultimately anything that returns a GraphQL-formatted response, such as a Federation router, is probably the API endpoint itself, and so this comment does not really apply as the response is being handled there, and not passed opaquely back to the client.

Reply:
Agree for a Federation router this sentence makes not sense but what about a CDN? 

Let's say you put your GraphQL server behind Cloudflare using persisted documents so everything that Cloudflare sees is REST (don't ask me why anyone would do that, I'm just a poor client developer but let's assume it's a thing). 

If Cloudflare detects that a botnet is using your endpoint, they're going to reply with:

```
429 Too many
Content-type: application/json

{
  "message": "oh no we detected a botnet",
  "..."
}
```

In the 2xx case, we're sure the source of the response is the GraphQL server but not in the other cases? At least this is my understanding of it.

Reply:
Correct.  I was just saying that whether the CDN "understands" GraphQL doesn't really make sense.  The CDN is a caching service.  A proxy or firewall is a relay service.  They don't parse GraphQL requests; if they did, such as in Apollo Federation, they are essentially the endpoint themselves and the GraphQL server is just responding to requests from the firewall.  The goal here is to just ensure that a 400 application/json response code from the GraphQL server can be distinguished from a 400 application/json response code from an intermediary server.

Reply:
>  I was just saying that whether the CDN "understands" GraphQL doesn't really make sense. 

The sentence is informative only. Just highlights that there might be non-GraphQL links in the chain, which you might forget if you're reading the spec from a client point of view. Agree it's technically not needed but I don't think it hurts?

> The goal here is to just ensure that a 400 application/json response code from the GraphQL server can be distinguished from a 400 application/json response code from an intermediary server.

Ah! This might be what I'm missing there. Is that possible based on just `content-type` + `status code`? In the general case, a CDN error could potentially clash with any GraphQL error response?

Reply:
Yeah, so like the `application/json` + 400 could be CloudFlare's response, or it could be a GraphQL response.  So we recommend that 200 is returned for `application/json` responses even in the case of an error.  Or, use `application/graphql-response+json` so that the response is known to be able to be parsed by the client, even in the even of a 400 error.

Reply:
Gotcha 👍 Makes sense. I think this is covered [here](https://graphql.github.io/graphql-over-http/draft/#sel-FANNLHCCBCOwDl2O)